### PR TITLE
[FEATURE] Schema: Introduce specVersion 4.0

### PR DESCRIPTION
**Pull request by d3xter666**
JIRA: CPOUI5FOUNDATION-835

### CODE REVIEW

File: lib/validation/schema/specVersion/kind/project/application.json [hunk 1]
```diff
@@ -77,25 +76,26 @@
 					"additionalProperties": true
 				}
 			}
-		},
+		},
```
Comment (flovogt) [inline lib/validation/schema/specVersion/kind/project/application.json:79]: trailing whitespace

Comment (d3xter666) [reply lib/validation/schema/specVersion/kind/project/application.json:79]: Thanks!

File: test/lib/validation/schema/__helper__/builder-bundleOptions.js [hunk 2]
```diff
@@ -13,7 +13,128 @@ export default {
...
+						"kind": "project",
+						"type": type,
+						"metadata": {
+							"name": "com.sap.ui5.test",
+							"copyright": "yes"
+						},
+						"builder": {
+							"bundles": [{
+								"bundleOptions": {
+									"debugMode": true
```
Comment (RandomByte) [inline test/lib/validation/schema/__helper__/builder-bundleOptions.js:53]:
```suggestion
									"usePredefineCalls": true
```

debugMode removal is already tested further down

File: test/lib/validation/schema/specVersion/kind/project/application.js [hunk 3]
```diff
@@ -46,7 +46,253 @@ test.after.always((t) => {
...
+									"resolve": true,
+									"sort": true,
+									"declareRawModules": false
+								},
+								{
+									"mode": "require",
+									"filters": [
+										"ui5loader-autoconfig.js"
+									],
+									"resolve": true
```
Comment (RandomByte) [inline test/lib/validation/schema/specVersion/kind/project/application.js:135]:
```suggestion
									"resolve": true,
									"async": false
```

Test the new property

File: test/lib/validation/schema/specVersion/kind/project/library.js [hunk 4]
```diff
@@ -46,7 +46,180 @@ test.after.always((t) => {
...
+									"resolve": true,
+									"sort": true,
+									"declareRawModules": false
+								},
+								{
+									"mode": "require",
+									"filters": [
+										"ui5loader-autoconfig.js"
+									],
+									"resolve": true
```
Comment (RandomByte) [inline test/lib/validation/schema/specVersion/kind/project/library.js:135]:
```suggestion
									"resolve": true,
									"async": false
```

**Review by RandomByte — CHANGES_REQUESTED** [review]

**Review by RandomByte — APPROVED** [review]

File: lib/validation/schema/specVersion/kind/project.json [hunk 5]
```diff
@@ -381,6 +396,120 @@
...
+				"defaultFileTypes": {
+					"type": "array",
+					"items": {
+						"type": "string"
+					}
+				},
+				"sections": {
+					"type": "array",
+					"if": {
+						"$comment": "Add async prop only if mode = 'require'",
```
Comment (matz3) [inline lib/validation/schema/specVersion/kind/project.json:416]: this constraint does not work. I suspect that the "then" section must not list all the other modes and only list "require" (see comment below).

It looks like there's a test missing to verify that the async flag is not allowed in other section modes.

File: lib/validation/schema/specVersion/kind/project.json [hunk 6]
```diff
@@ -381,6 +396,120 @@
...
+						"items": {
+							"type": "object",
+							"additionalProperties": false,
+							"required": ["mode", "filters"],
+							"properties": {
+								"name": {
+									"type": "string"
+								},
+								"mode": {
+									"enum": ["raw", "preload", "require", "provided", "bundleInfo", "depCache"]
```
Comment (matz3) [inline lib/validation/schema/specVersion/kind/project.json:434]: here, only "require" should be listed, as only then the "async" flag is supported.

File: lib/validation/schema/specVersion/kind/project.json (hunk 5 above)
Comment (d3xter666) [reply lib/validation/schema/specVersion/kind/project.json:416]: Fixed with: https://github.com/SAP/ui5-project/pull/751

File: lib/validation/schema/specVersion/kind/project.json (hunk 6 above)
Comment (d3xter666) [reply lib/validation/schema/specVersion/kind/project.json:434]: Fixed with: https://github.com/SAP/ui5-project/pull/751

**Pull request merged**